src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,1305 @@
+// can be driven on (not an obstacle).
+bool vehicle::autodrive_controller::check_drivable( tripoint pt ) const
+{
+    const map &here = get_map();
+
+    // check if another vehicle is there; tiles occupied by the current
+    // vehicle are evidently drivable
+    const optional_vpart_position ovp = here.veh_at( pt );
+    if( ovp ) {
+        // Known corner case: some furniture can be driven over, but will collide with
+        // wheel parts; if the vehicle starts over such a furniture we'll mark that tile
+        // as safe and may collide with it by turning; however if we mark it unsafe
+        // we'll have no viable paths away from the starting point.
+        return &ovp->vehicle() == &driven_veh;
+    }
+
+    const tripoint_abs_ms pt_abs( here.getabs( pt ) );
+    const tripoint_abs_omt pt_omt = project_to<coords::omt>( pt_abs );
+    // only check visibility for the current OMT, we'll check other OMTs when
+    // we reach them
+    if( pt_omt == data.current_omt ) {
+        // driver must see the tile or have seen it before in order to plan a route over it
+        if( !driver.sees( pt ) ) {
+            if( !driver.is_avatar() ) {
+                return false;
+            } else if( driver.as_avatar()->get_memorized_tile( pt_abs.raw() ) == mm_submap::default_tile ) {
+                // apparently open air doesn't get memorized, so pretend it is or else
+                // we can't fly helicopters due to the many unseen tiles behind the driver
+                if( !( data.air_ok && here.ter( pt ) == t_open_air ) ) {
+                    return false;
+                }
+            }
+        }
+    }
+
+    // check for creatures
+    // TODO: padding around monsters
+    Creature *critter = g->critter_at( pt, true );
+    if( critter && driver.sees( *critter ) ) {
+        return false;
+    }
+
+    // check for furniture that hinders movement; furniture with 0 move cost
+    // can be driven on
+    const furn_id furniture = here.furn( pt );
+    if( furniture != f_null && furniture.obj().movecost != 0 ) {
+        return false;
+    }
+
+    const ter_id terrain = here.ter( pt );
+    if( terrain == t_null ) {
+        return false;
+    }
+    // open air is an obstacle to non-flying vehicles; it is drivable
+    // for flying vehicles
+    if( terrain == t_open_air ) {
+        return data.air_ok;
+    }
+    const ter_t &terrain_type = terrain.obj();
+    // watercraft can drive on water
+    if( data.water_ok && terrain_type.has_flag( TFLAG_SWIMMABLE ) ) {
+        return true;
+    }
+    // remaining checks are for land-based navigation
+    if( !data.land_ok ) {
+        return false;
+    }
+    if( terrain_type.movecost <= 0 ) {
+        // walls and other impassable terrain
+        return false;
+    } else if( terrain_type.movecost == 2 || terrain_type.has_flag( "NOCOLLIDE" ) ) {
+        // terrain with neutral move cost or tagged with NOCOLLIDE will never cause
+        // collisions
```

Comment:
```suggestion
    } else if( terrain_type.movecost == 2 || terrain_type.has_flag( "NOCOLLIDE" ) 
        || ( terrain_type.has_flag( "FLAT" ) && terrain_type.has_flag( "ROAD" ) && terrain_type.movecost == 1 ) ) {
        // terrain with neutral move cost or tagged with NOCOLLIDE will never cause
        // collisions
```
This may need reformatting; the above is to handle some _open_ gates and similar that have move_cost 1; see terrain-doors.json, terrain-fences-gates.json, and terrain-migo.json. (Admittedly, if planning without knowledge of the state of said gates currently, it may be best to be conservative. Again, this may be a use case for "OK but increased cost".)

Reply:
Note that if none of the if branches are taken the default is to return true.
I don't think we need a special case for move_cost=1, but let me check that in-game.

Reply:
Ok, I used the map editor to place an open metal gate in front of the vehicle. Autodrive drove around it.
I tried manually driving through and the vehicle crashed into it. So I'd say it's working correctly.

Reply:
OK. I think that may be due to vehicle width, actually.

Reply:
The obstacle detection logic in check_drivable() closely matches what's in vehicle::part_collision()
Added a comment about keeping these in sync. It would be better to refactor and share a common function, but that's for a later PR.